metadata/src/main/java/org/apache/kafka/controller/ReplicationControlManager.java

```diff
@@ -541,6 +559,41 @@ static void validateNewTopicNames(Map<String, ApiError> topicErrors,
         return configChanges;
     }
 
+    ControllerResult<DeleteTopicsResponseData> deleteTopics(Map<String, Uuid> nameToId) {
+        DeleteTopicsResponseData result = new DeleteTopicsResponseData();
+        List<ApiMessageAndVersion> records = new ArrayList<>();
+        for (Entry<String, Uuid> entry : nameToId.entrySet()) {
+            ApiError error = deleteTopic(entry.getKey(), entry.getValue(), records);
+            result.responses().add(new DeletableTopicResult().
+                setName(entry.getKey()).
+                setTopicId(entry.getValue()).
+                setErrorCode(error.error().code()).
+                setErrorMessage(error.message()));
+        }
+        return new ControllerResult<>(records, result);
+    }
+
+    ApiError deleteTopic(String name,
+                         Uuid providedId,
+                         List<ApiMessageAndVersion> records) {
+        Uuid realId = topicsByName.get(name);
+        if (realId == null) {
+            return new ApiError(UNKNOWN_TOPIC_OR_PARTITION,
+                "Unable to locate the provided topic name.");
+        }
+        if (!providedId.equals(Uuid.ZERO_UUID) && !providedId.equals(realId)) {
+            return new ApiError(UNKNOWN_TOPIC_ID,
+                "The provided topic ID does not match the provided topic name.");
+        }
+        TopicControlInfo topic = topics.get(realId);
+        if (topic == null) {
+            return new ApiError(UNKNOWN_TOPIC_OR_PARTITION, "Unable to locate topic id.");
```

Comment:
Is this a server-side bug? the topic id exists in `topicsByName` but there is no `TopicControlInfo`.

Reply:
this is out of date